### PR TITLE
cp: move copy_directory() to its own module

### PR DESCRIPTION
**Pull request by jfinkels**
This pull request refactors `copy_directory()` and related helper functions into its own new module `copydir.rs`. This commit also adds some additional structs and helper functions to make the code easier to read and maintain. This commit does not change the behavior of the `copy_directory()` function, only the organization of the code.

I am proposing this change because I felt while working on pull request #3894 that it was harder than necessary for me to read and understand the code.

### CODE REVIEW

File: src/uu/cp/src/copydir.rs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        match direntry_result {
+            Ok(direntry) => {
+                let entry = Entry::new(&context, &direntry)?;
+                copy_direntry(
+                    entry,
+                    options,
+                    symlinked_files,
+                    preserve_hard_links,
+                    &mut hard_links,
+                )?;
```
Comment (tertsdiepraam) [inline src/uu/cp/src/copydir.rs:300]: Is it okay to bubble the errors up here? Shouldn't they go to the `show_error!` on lin 303?

**Comment by tertsdiepraam** [comment]
Looks good! I just have the one comment above but otherwise ready to merge. Could you separate the moving around and refactoring next time, so it's easier to review?

**Comment by jfinkels** [comment]
> Could you separate the moving around and refactoring next time, so it's easier to review?

Yes, absolutely.

File: src/uu/cp/src/copydir.rs (hunk 1 above)
Comment (jfinkels) [reply src/uu/cp/src/copydir.rs:300]: I was matching the previous behavior. See the main loop in the old code, starting with:
```

    for path in WalkDir::new(root)
        .same_file_system(options.one_file_system)
        .follow_links(options.dereference)
    {
```
There's a bunch of `?` and `return Err(...)` that will terminate the loop immediately and exit the `copy_directory()` function with an error. I didn't check whether this is the right behavior when compared with GNU cp, but it seemed to be what the old code was doing. The `show_error!()` on line 303 corresponds to the `or_continue!()` macro in the old code.

**Comment by jfinkels** [comment]
I rebased and updated this branch so that one commit moves the `copy_directory()` function as-is to a new module, and the next commit breaks up the `copy_directory()` function into smaller helper functions.

**Comment by sylvestre** [comment]
Conflicting again after the merge of your other PR ;)

**Pull request merged**